pydantic/types.py

```diff
@@ -1759,7 +1760,16 @@ class MyModel(BaseModel):
 
     @classmethod
     def __get_pydantic_core_schema__(cls, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
-        return core_schema.with_info_plain_validator_function(cls._validate)
+        return core_schema.with_info_after_validator_function(
+            function=cls._validate,
+            schema=core_schema.union_schema(
+                [
+                    core_schema.str_schema(pattern=byte_string_pattern),
+                    core_schema.int_schema(ge=0),
+                ]
+            ),
+            serialization=core_schema.plain_serializer_function_ser_schema(function=cls._serialize),
```

Comment:
```suggestion
                        serialization=core_schema.plain_serializer_function_ser_schema(lambda v: int(v), return_schema=core_schema.int_schema(ge=0)),
```

I think you should be able to use something as simple as this to get the serialization schema (and logic) correct, and then you can omit the below `_serialize` function + modify the test. The `validation` and `serialization` schemas should differ in the test.

Reply:
ah, of course, that makes sense now!

made the change but the output is still the same for both.

Reply:
ah, ok, that makes sense now.